File: 큐와 덱/덱.py
# 10866번 : 덱
import sys

deque = []
N = int(sys.stdin.readline())

for i in range(N):
    raw_input = sys.stdin.readline().rstrip()
    try:
        prompt, num = raw_input.split()
        num = int(num)
    except:
        prompt = raw_input

    if (prompt == "push_front"):
        deque.insert(0, num)
    elif (prompt == "push_back"):
        deque.append(num)
    elif (prompt == "pop_front"):
        if (deque):
            print(deque[0])
            del deque[0]
        else:
            print(-1)
    elif (prompt == "pop_back"):
        if (deque):
            print(deque[-1])
            del deque[-1]
        else:
            print(-1)
    elif (prompt == "size"):
        print(len(deque))
    elif (prompt == "empty"):
        if (deque):
            print(0)
        else:
            print(1)
    elif (prompt == "front"):
        if (deque):
            print(deque[0])
        else:
            print(-1)
    elif (prompt == "back"):
        if (deque):
            print(deque[-1])
        else:
            print(-1)

File: 알고리즘 스터디 문제들/Part 2 - 자료구조/문제/덱.cpp
#include <iostream>
#include <string>
using namespace std;

struct deque
{
    int arr[20'000 + 1];
    int head = 10'000;
    int tail = 10'000;
};

void push_front(deque &D, int X)
{
    D.arr[--D.head] = X;
}

void push_back(deque &D, int X)
{
    D.arr[D.tail++] = X;
}

bool empty(deque &D)
{
    return (D.tail - D.head) == 0;
}

int pop_front(deque &D)
{
    if (empty(D))
        return -1;
    else
        return D.arr[D.head++];
}

int pop_back(deque &D)
{
    if (empty(D))
        return -1;
    else
        return D.arr[--D.tail];
}

int size(deque &D)
{
    return (D.tail - D.head);
}

int front(deque &D)
{
    if (empty(D))
        return -1;
    else
        return D.arr[D.head];
}

int back(deque &D)
{
    if (empty(D))
        return -1;
    else
        return D.arr[D.tail - 1];
}

// Class 구현도 가능하지만 일단은 struct만으로 구현해보자.
int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);

    size_t N;
    cin >> N;

    deque D;
    string input;
    for (size_t i = 0; i < N; i++)
    {
        cin >> input;

        if (input == "push_front")
        {
            int tmp;
            cin >> tmp;
            push_front(D, tmp);
        }
        else if (input == "push_back")
        {
            int tmp;
            cin >> tmp;
            push_back(D, tmp);
        }
        else if (input == "pop_front")
            cout << pop_front(D) << '\n';
        else if (input == "pop_back")
            cout << pop_back(D) << '\n';
        else if (input == "size")
            cout << size(D) << '\n';
        else if (input == "empty")
            cout << empty(D) << '\n';
        else if (input == "front")
            cout << front(D) << '\n';
        else if (input == "back")
            cout << back(D) << '\n';
    }
}
